R/app.R

```diff
@@ -496,3 +506,17 @@ knit_print.reactive <- function(x, ..., inline = FALSE) {
     x()
   }), inline = inline)
 }
+
+#' Run code after app has finished
+#'
+#' This function allows you to register a callback function that is invoked
+#' after your app has finished running (all sessions have been disconnected).
+#'
+#' @param fun A function that will be called after the app has finished running.
+#'
+#' @examples
+#' onStop(function() { print("goodbye") })
+#' @export
+onStop <- function(fun) {
+  shinyOptions(onStop = fun)
```

Comment:
I liked the fact that this mirrors `onStart` and it stays simple, yet powerful. I also wonder if allowing multiple `onStop` calls won't make it easy for app authors to use it for session-specific things (i.e. calling `onStop` in reactives or observers, since it wouldn't always fail -- I guess we could enforce this if we wanted to). Having just one callback possible also both forces the app author to think about order and shifts that burden to them. 

I guess that what I'm getting at is why multiple `onStop` calls would be useful if this is a global thing, rather than session-specific. And I'm pretty sure there's one answer to rule them all: modules, right?

Assuming we go your way, is the best way to implement this to use a list in `.globals`? What's the API you imagine for this + cancellation of individual callbacks? If we didn't have cancellation, it could stay the same, except that multiple calls would result in multiple callbacks, instead of the last one winning out. With cancellation in mind, the API that makes most sense to me is a la `scheduleTask()` in pool. Applied to this, the return value of an `onStop` call would a function that cancels the callback originally registered:

```r
cancel <- onStop(function(){}) # this registers the callback
cancel() # this cancels the callback
```